libraries/chain/controller.cpp

```diff
@@ -3198,15 +3265,39 @@ struct controller_impl {
                report_block_header_diff(*b, ab.header());
 
                // this implicitly asserts that all header fields (less the signature) are identical
-               EOS_ASSERT(producer_block_id == ab.id(), block_validate_exception, "Block ID does not match",
+               EOS_ASSERT(producer_block_id == ab.id(), block_validate_exception, "Block ID does not match, ${producer_block_id} != ${validator_block_id}",
                           ("producer_block_id", producer_block_id)("validator_block_id", ab.id()));
             }
 
+            // verify received finality digest in action_mroot is the same as the actual one
+            if constexpr (std::is_same_v<BSP, block_state_ptr>) {
+               // For proper IF blocks that do not have an associated Finality Tree defined,
+               // its finality_mroot is empty
+               digest_type actual_finality_mroot{};
+               if (!bsp->core.is_genesis_block_num(bsp->core.final_on_strong_qc_block_num)) {
+                  actual_finality_mroot = bsp->get_validation_mroot(bsp->core.final_on_strong_qc_block_num);
+               }
+
+               EOS_ASSERT(bsp->finality_mroot() == actual_finality_mroot,
+                  block_validate_exception,
+                  "finality_mroot does not match, received finality_mroot: ${r} != actual_finality_mroot: ${a}",
+                  ("r", bsp->finality_mroot())("a", actual_finality_mroot));
+            }
+
             if( !use_bsp_cached ) {
                bsp->set_trxs_metas( ab.extract_trx_metas(), !skip_auth_checks );
             }
             // create completed_block with the existing block_state as we just verified it is the same as assembled_block
-            pending->_block_stage = completed_block{ block_handle{bsp} };
+            if constexpr (std::is_same_v<BSP, block_state_legacy_ptr>) {
+               // Store action_receipt_digests in completed_block for
+               // calculating action_mroot in the Savanna way during
+               // Legacy to Savanna transition
+               assert(std::holds_alternative<assembled_block::assembled_block_legacy>(ab.v));
+               auto legacy_ab = std::get<assembled_block::assembled_block_legacy>(ab.v);
+               pending->_block_stage = completed_block{ block_handle{bsp}, legacy_ab.action_receipt_digests };
+            } else {
+               pending->_block_stage = completed_block{ block_handle{bsp} };
+            }
```

Comment:
This is awkward. What about having  both `assembled_block` versions have a member get_action_receipt_digests()`, so you can replace with:

```suggestion
               pending->_block_stage = completed_block{ block_handle{bsp}, ab.get_action_receipt_digests() };
```